### ISSUE TITLE
Feature request: make object name configurable / enable partitioning

### ISSUE THREAD

**Issue by tsehrer** [ibm-messaging/kafka-connect-ibmcos-sink#3]
I'd like to request a feature to make an object name configurable and/or the ability to specify partitions. The S3 Connector for Kafka Connect has this feature and it's been enormously useful for pulling specific data out of COS with Cloud SQL, among other uses.

**Comment by prestona** [comment]
Hi @tsehrer - at the moment the object name is fixed and is built from the topic name, partition number, first record offset, and last record offset. Are you asking for the flexibility to change the order of these fields and their delimiters? Or are you also asking to include other information in the object name (and if so, what? :smile:)

**Comment by prestona** [comment]
Could you also elaborate a bit more about what you mean by "ability to specify partitions"? Are you looking to only consume from certain partitions within a Kafka topic? Or something else?

**Comment by mimaison** [comment]
@tsehrer Can you elaborate on your requirements? We're happy to evolve this connector but would like to understand your needs clearly before we start making changes.

**Comment by mtsargent** [comment]
I'm not sure if this is the same item being requested in this issue, but I came across this repository while searching for a way to take a single Kafka topic and write data to multiple root directories in COS based on the contents of a Kafka message. A sort of partitioning based message contents.

For example, suppose a topic had these messages:
```
{"messageType":"TypeA", "groupId":"1234", "payload":"P1"}
{"messageType":"TypeA", "groupId":"1234", "payload":"P2"}
{"messageType":"TypeA", "groupId":"9876", "payload":"P3"}
{"messageType":"TypeA", "groupId":"9876", "payload":"P4"}
{"messageType":"TypeB", "groupId":"9876", "payload":"P5"}
```

I would like the ability to group data on COS split by groupId and messageType (preferably by both, but grouping by one field could also be useful). I would want data to be grouped on COS for `TypeA/1234`, `TypeA/9876`, and `TypeB/9876`.

This sounds similar to the Field Partitioner referenced here: https://docs.confluent.io/current/connect/kafka-connect-s3/index.html. Although, I am not sure if that Field Partitioner allows you to specify multiple fields for partitioning.

If you would like, I could move this to a separate issue. I thought it sounded similar enough to this open issue that I would try here first.

**Comment by prestona** [comment]
Interesting. When this issue was first raised, I wondered about making object name configurable based on a template. I was thinking of things like: topic name, partition, offset range, and timestamp as values that could be substituted into the template.

It sounds like you would also want the template to include properties of the JSON message payload? These would need to be combined with values that ensured each object is unique (such as partition number and offset range). So you might end up with something like:
```
TypeA/1234/1/0000000000017805-0000000000017809
```
Would this kind of naming scheme work for your intended use of the connector?

**Comment by mtsargent** [comment]
I will have to think about if partition and offset in the path makes things strange at all. Once data is out of Kafka we do not care about partition or offset (just the fields that were part of the json message).

I suppose we could try ignoring the partition and offset when reading in data from COS.

**Comment by prestona** [comment]
Some background here: COS doesn't really have the concept of appending to an existing object. So the connector reads records from Kafka, batches them together based on some criteria, and then writes the batch of records into COS as an object. If the connector didn't include something like the offsets, then each successive write to the same object would overwrite whatever was previously in the object.

The _unique_ part of the object name doesn't have to be a range of offsets, but to ensure that the data is written without loss or duplication it needs to be something derived from the Kafka records that make up the batch. For example message timestamps for the first and last messages would probably also be possible.

**Comment by mtsargent** [comment]
Gotcha. Thanks for the clarification. I was thinking the example with offsets was stopping at the "folder" before any files would be written out.

In that case, I think partition and offsets/timesets seem like they should work.

**Comment by jhughes24816** [comment]
From reading through the conversation it seems this has been resolved and should have been closed a long time ago. I will close it shortly. If there are follow up questions please raise a new issue and reference the conversation here. Thanks.
